### DISCUSSION TITLE
coursera: Nand2Tetris - CS%E8%87%AA%E5%AD%A6%E6%8C%87%E5%8D%97

### DISCUSSION THREAD

**Discussion by giscus** [PKUFlyingPig/cs-self-learning#161] [Announcements]
# coursera: Nand2Tetris - CS%E8%87%AA%E5%AD%A6%E6%8C%87%E5%8D%97

CS自学指南

https://csdiy.wiki/%E4%BD%93%E7%B3%BB%E7%BB%93%E6%9E%84/N2T/

**Comment by EveryThinUknow** [comment]
课程教材不见了！

**Reply by socialismchuanren** [reply]
资源汇总里面有

**Comment by AllenWrong** [comment]
https://github.com/AllenWrong/nand2tetris  分享我用Java写的，欢迎PR

**Comment by lng205** [comment]
在 https://github.com/lng205/Nand2Tetris 分享了基于2021版教材的笔记、作业及分析，对硬件部分知识做了更深入的补充。

**Reply by lng205** [reply]
项目内提供了课程相关资源的链接，需要离线版资源可以私聊联系我。

**Reply by dengyunsheng250** [reply]
你好，请问怎么联系？

**Reply by lng205** [reply]
可发邮件到主页的邮箱，我这有教材pdf

**Comment by huijian222** [comment]
[2021新版本的课本pdf](https://dokumen.pub/the-elements-of-computing-systems-building-a-modern-computer-from-first-principles-2nbsped-2020002671-9780262539807.html)

**Reply by YinZai-SHA** [reply]
非常感谢！

**Comment by fkgkdfgy** [comment]
https://github.com/fkgkdfgy/from_nand_2_tetris 分享一个Python版本的作业。

**Comment by Eleven-6** [comment]
如果英语水平很烂的话（比如我），一个月学不完！！！

**Reply by AuroBreeze** [reply]
可以买《计算机系统要素》这本书，没啥问题，

**Comment by LysanderT** [comment]
https://github.com/LysanderT/Nand2Tetris/tree/master
（含教材）

发现根本没人用c++写。。。。

**Reply by LysanderT** [reply]
自顶一个

**Comment by NULL7272** [comment]
project03ing(dog

**Comment by zhangeecs** [comment]
想问下大家，这门课在Coursera有必要申请助学金吗？还是说旁听就行了

**Reply by JosephJostar0** [reply]
可以只旁听，旁听只是无法提交Project，其他内容没有缺失，而课程非常友好的提供了Test Scripts帮你检查Project的正误  
[Test Scripts 详见课程官网](https://www.nand2tetris.org/software)

**Comment by hongkuancn** [comment]
感谢楼主的推荐，收获很大！用时大概80小时。贴一个go的实现，欢迎讨论

https://github.com/hongkuancn/nand2tetris

**Comment by sbwcwso** [comment]
个人学习资料汇总及笔记:  [nand2tetris 自学笔记](https://sbwcwso.github.io/CSStudy/#/page/nand2tetris)
个人 Project 仓库参考: [project 仓库 - python 实现](https://github.com/sbwcwso/cs61b-sp18)

这门课程的所有内容和 Project 我基本上都学习并做了一遍，感觉收获还是蛮多的。感谢作者提供相关课程的信息资源。

同时如果学习过程中碰到具体疑问或者单纯只是想交流学习经验和心得，欢迎到我的相关仓库中进行讨论。

**Comment by kazusa3e** [comment]
贴一个自己的实现：https://github.com/kazusa3e/coursera-nand2tetris

但是只做了硬件部分

**Comment by HaleOIC** [comment]
https://github.com/HaleOIC/Nand2Tetris
贴一个我的实现，本来是打算用rust的，但是coursera的评测机没有安装rustc，后面改成了java实现的版本，欢迎PR，代码已经全部通过coursera的评测系统，大体上是满分实现（如果coursera上面评测是complete的话）

**Comment by Kidding0204** [comment]
注意两位老师毕竟来自非英语母语国家，所以英语口语水平“较为一般”。我平常听播客的突然听这个实在是有些接受不了，如果你也感到有些折磨的话，推荐用阅读教材替代视频（视频讲解确实简单些，书本更具深度与广度，但大体内容是一样的）。

**Comment by MandyZhou0314** [comment]
贴一下part1的作业python实现，欢迎学习指教~~
https://github.com/MandyZhou0314/nand2tetris_repo/tree/main

**Reply by kawayee** [reply]
project6写的不行，claude给的点评：
1. Symbol_table 的 st 是类变量，不是实例变量；
2. Parser.__init__(self, raw_rows=[]) 用可变对象作默认参数；
3. split("=") 和 split(";") 应该限制 maxsplit=1；
4. instructionType() 用 in 判断不严谨；
5. 没有asm后缀校验；
6. project太小，不应该放在四个文件夹里，会污染path；
然后我发现的，lecture-ppt里有api文档，不用自己决定函数，first-pass和second-pass的实现应该在main里，不应该放在parser里。

**Comment by louis-chiu** [comment]
請問這門課足以滿足作業系統的計算機體系結構的先修要求嗎？

**Reply by isxiaohe** [reply]
足够！但是更推荐cs61c！

**Reply by louis-chiu** [reply]
目前是走 web backend，主要想補 OS 的觀念，所以才想說計算機體系結構以快速走過觀念為主，你會推薦 Nand2teris  I 和 II 都要完成才繼續學操作系統嗎？

**Reply by HaleOIC** [reply]
II不用太看吧，主要是讲编译的，看看I直接就可以上os了

**Comment by Ira569** [comment]
很棒的课，记得自己好像把硬件部分全做完了，但可惜当时没有编程基础，一到软件完全写不出来

**Comment by XiaoWan-xi** [comment]
学体系结构还可以推荐steam上一个游戏：turing complete(图灵完备)  非常合适

**Comment by yykina** [comment]
中文教材67页图4.3有两个“a=0”😄

**Comment by Jing-Yike** [comment]
这课为啥没法免费上啊

**Reply by sbwcwso** [reply]
不要证书就是免费吧。

**Reply by Jing-Yike** [reply]
但是现在enroll了之后好像没有audit的选项了，只能听完第一模块，然后就要付钱了

**Reply by sbwcwso** [reply]
> 但是现在enroll了之后好像没有audit的选项了，只能听完第一模块，然后就要付钱了

重新注册个账号就可以了吧。

**Reply by Jing-Yike** [reply]
是不是这个课现在取消掉了免费旁听啊

**Reply by dbxxx-ac** [reply]
好像是的，我也找不到 audit 选项，目前打算看书学习了

**Reply by Jing-Yike** [reply]
我看B站搬运的前几年的视频学习的，也问题不大

**Comment by Mock-Canine** [comment]
Part I 部分的代码，最终的project6使用python实现，后续会跟进软件部分 https://github.com/Mock-Canine/Nand2Tetris/tree/main

**Comment by hsh-yellow3** [comment]
真正开始认真学的第一门课，学着感觉真的是能够重新筑基，计算机没有黑魔法 哈哈 XD。
资源：b站有个其他UP主的在19年的搬运，并且我亲自在Coursera上面看到的课和19年的搬运没什么区别。此外program也重新改成了线上的IDE，如果要保存进度在线上的网站上面下载了。一开始学可能会觉得“讲的这么精简，好像有些该讲的没讲？有些好像在其他学科学过？但是这里这么不解释了？”
但是学着学着就可以发现这两个老师在课程设计方面的小巧思，可以说教学功力非常深厚了

网站：https://www.bilibili.com/video/BV1g4411A7iU?spm_id_from=333.788.player.switch&vd_source=bc97f52cd8edff725fb81cd1ea86106b&p=31 
但是字幕是纯英文可能需要开一个沉浸式翻译（其他插件做不到一次性翻译字幕）
b站视频的简介有视频up的简书，可以做个参考

网页IDE项目地址如下：注意小齿轮可以选择下载内容到本地
https://nand2tetris.github.io/web-ide/chip/

**Reply by hsh-yellow3** [reply]
补充（暂时不知道用处在哪）：
https://www.nand2tetris.org/demos
